File: backend/app/rules_engine/combos.py
from dataclasses import dataclass
from enum import StrEnum

from app.rules_engine.cards import TWO_RANK, Card, card_rank


class ComboType(StrEnum):
    SINGLE = "single"
    PAIR = "pair"
    TRIPLE = "triple"
    STRAIGHT = "straight"
    QUAD = "quad"
    CONSECUTIVE_PAIRS = "consecutive_pairs"


BOMB_TYPES = (ComboType.QUAD, ComboType.CONSECUTIVE_PAIRS)

# Exact chop table: pile length (of rank-2 cards) -> allowed (type, length) bombs.
TWO_CHOPS: dict[int, set[tuple[ComboType, int]]] = {
    1: {(ComboType.QUAD, 4), (ComboType.CONSECUTIVE_PAIRS, 3)},
    2: {(ComboType.CONSECUTIVE_PAIRS, 4)},
    3: {(ComboType.CONSECUTIVE_PAIRS, 5)},
}


@dataclass(frozen=True)
class Combo:
    type: ComboType
    cards: tuple[Card, ...]
    length: int
    high: Card
    is_bomb: bool


def _is_consecutive(ranks: list[int]) -> bool:
    return all(ranks[i + 1] - ranks[i] == 1 for i in range(len(ranks) - 1))


def parse_combo(cards: list[Card]) -> Combo | None:
    if not cards or len(set(cards)) != len(cards):
        return None

    sorted_cards = sorted(cards)
    rank_groups: dict[int, list[Card]] = {}
    for c in sorted_cards:
        rank_groups.setdefault(card_rank(c), []).append(c)

    group_sizes = [len(g) for g in rank_groups.values()]
    unique_ranks = sorted(rank_groups.keys())
    n = len(sorted_cards)

    combo_type: ComboType
    length: int

    if n == 1:
        combo_type, length = ComboType.SINGLE, 1
    elif n == 2 and len(rank_groups) == 1:
        combo_type, length = ComboType.PAIR, 2
    elif n == 3 and len(rank_groups) == 1:
        combo_type, length = ComboType.TRIPLE, 3
    elif n == 4 and len(rank_groups) == 1:
        combo_type, length = ComboType.QUAD, 4
    elif (
        all(size == 2 for size in group_sizes)
        and len(rank_groups) >= 3
        and _is_consecutive(unique_ranks)
        and TWO_RANK not in unique_ranks
    ):
        combo_type, length = ComboType.CONSECUTIVE_PAIRS, len(rank_groups)
    elif (
        all(size == 1 for size in group_sizes)
        and n >= 3
        and _is_consecutive(unique_ranks)
        and TWO_RANK not in unique_ranks
    ):
        combo_type, length = ComboType.STRAIGHT, n
    else:
        return None

    return Combo(
        type=combo_type,
        cards=tuple(sorted_cards),
        length=length,
        high=max(sorted_cards),
        is_bomb=combo_type in BOMB_TYPES,
    )


def beats(candidate: Combo, pile: Combo) -> bool:
    pile_is_all_twos = pile.type in (
        ComboType.SINGLE,
        ComboType.PAIR,
        ComboType.TRIPLE,
    ) and card_rank(pile.cards[0]) == TWO_RANK

    if pile_is_all_twos and candidate.is_bomb:
        allowed = TWO_CHOPS.get(pile.length, set())
        return (candidate.type, candidate.length) in allowed

    if pile.is_bomb and candidate.is_bomb:
        if (
            pile.type == ComboType.QUAD
            and candidate.type == ComboType.CONSECUTIVE_PAIRS
            and candidate.length == 3
        ):
            return True
        if candidate.type == pile.type and candidate.length == pile.length:
            return candidate.high > pile.high
        return False

    if candidate.type == pile.type and candidate.length == pile.length:
        return candidate.high > pile.high
    return False

File: frontend/src/lib/rules-engine/combos.ts
import { TWO_RANK, cardRank, type Card } from "./cards";

export type ComboType = "single" | "pair" | "triple" | "straight" | "quad" | "consecutive_pairs";

export interface Combo {
  type: ComboType;
  cards: Card[];
  length: number;
  high: Card;
  isBomb: boolean;
}

const BOMB_TYPES: ComboType[] = ["quad", "consecutive_pairs"];

// Exact chop table: pile length (of rank-2 cards) -> allowed [type, length] bombs.
const TWO_CHOPS: Record<number, [ComboType, number][]> = {
  1: [
    ["quad", 4],
    ["consecutive_pairs", 3],
  ],
  2: [["consecutive_pairs", 4]],
  3: [["consecutive_pairs", 5]],
};

function isConsecutive(ranks: number[]): boolean {
  return ranks.every((rank, i) => i === 0 || rank - ranks[i - 1] === 1);
}

export function parseCombo(cards: Card[]): Combo | null {
  if (cards.length === 0 || new Set(cards).size !== cards.length) {
    return null;
  }

  const sortedCards = [...cards].sort((a, b) => a - b);
  const rankGroups = new Map<number, Card[]>();
  for (const card of sortedCards) {
    const rank = cardRank(card);
    const group = rankGroups.get(rank);
    if (group) {
      group.push(card);
    } else {
      rankGroups.set(rank, [card]);
    }
  }

  const groupSizes = [...rankGroups.values()].map((g) => g.length);
  const uniqueRanks = [...rankGroups.keys()].sort((a, b) => a - b);
  const n = sortedCards.length;

  let type: ComboType;
  let length: number;

  if (n === 1) {
    type = "single";
    length = 1;
  } else if (n === 2 && rankGroups.size === 1) {
    type = "pair";
    length = 2;
  } else if (n === 3 && rankGroups.size === 1) {
    type = "triple";
    length = 3;
  } else if (n === 4 && rankGroups.size === 1) {
    type = "quad";
    length = 4;
  } else if (
    groupSizes.every((size) => size === 2) &&
    rankGroups.size >= 3 &&
    isConsecutive(uniqueRanks) &&
    !uniqueRanks.includes(TWO_RANK)
  ) {
    type = "consecutive_pairs";
    length = rankGroups.size;
  } else if (
    groupSizes.every((size) => size === 1) &&
    n >= 3 &&
    isConsecutive(uniqueRanks) &&
    !uniqueRanks.includes(TWO_RANK)
  ) {
    type = "straight";
    length = n;
  } else {
    return null;
  }

  return {
    type,
    cards: sortedCards,
    length,
    high: Math.max(...sortedCards),
    isBomb: BOMB_TYPES.includes(type),
  };
}

export function beats(candidate: Combo, pile: Combo): boolean {
  const pileIsAllTwos =
    (pile.type === "single" || pile.type === "pair" || pile.type === "triple") &&
    cardRank(pile.cards[0]) === TWO_RANK;

  if (pileIsAllTwos && candidate.isBomb) {
    const allowed = TWO_CHOPS[pile.length] ?? [];
    return allowed.some(([type, length]) => candidate.type === type && candidate.length === length);
  }

  if (pile.isBomb && candidate.isBomb) {
    if (pile.type === "quad" && candidate.type === "consecutive_pairs" && candidate.length === 3) {
      return true;
    }
    if (candidate.type === pile.type && candidate.length === pile.length) {
      return candidate.high > pile.high;
    }
    return false;
  }

  if (candidate.type === pile.type && candidate.length === pile.length) {
    return candidate.high > pile.high;
  }
  return false;
}
